### PR TITLE
Update ONNXIFI documentation

### PR DESCRIPTION
**Pull request by Maratyszcza**
Reflect changed in API before ONNXIFI 1.0

### CODE REVIEW

File: docs/ONNXIFI.md [hunk 1]
```diff
@@ -20,13 +20,19 @@ We propose a cross-platform API for loading and executing ONNX graphs on optimiz
...
-6. When done using the model, release the model graph(s) with `onnxReleaseGraph`, then release the backend with `onnxReleaseBackend`
+5. Call `onnxSetGraphIO` to set locations and shapes for inputs and outputs of a graph.
+6. Initialize an `inputFence` structure of type `onnxMemoryFenceV1`: set `tag` to `ONNXIFI_TAG_MEMORY_FENCE_V1`, `type` to `ONNXIFI_SYNCHRONIZATION_EVENT`, and call `onnxInitEvent` to initiaze the `event` member.
+7. Initialize an `outputFence` structure of type `onnxMemoryFenceV1`: set `tag` to `ONNXIFI_TAG_MEMORY_FENCE_V1`, `type` to `ONNXIFI_SYNCHRONIZATION_EVENT`, and `event` to null.
+8. Call `onnxRunGraph` with the initialized `inputFence` and `outputFence` structures to enable execution of the graph. The call to `onnxRunGraph` will populate `event` member of the `outputFence` with a newly created event object, asynchronously execute the graph once `inputFence`'s `event` is signalled, and then signal the `outputFence`'s `event`.
+9. Call `onnxSignalEvent` with `event` member of `inputFence` to signal to the backend that the inputs are ready to be consumed.
```
[the remark this answers is not in the document]
Comment (yinghai) [reply docs/ONNXIFI.md]: The common trap is inside 8, we will probably call wait on `input event` and on CPU, I've seen implementation of events with conditional_variable. Have 9 after 8 will hang the process.

[the remark this answers is not in the document]
Comment (Maratyszcza) [reply docs/ONNXIFI.md]: 9 after 8 with condvar should work fine, so long as condvar wait happens on a different thread

[the remark this answers is not in the document]
Comment (yinghai) [reply docs/ONNXIFI.md]: Yeah, this is another common trap...

[the remark this answers is not in the document]
Comment (rdzhabarov) [reply docs/ONNXIFI.md]: @yinghai Currently, 8 and 9 are swapped in the pytorch/C2 implementation, right?

[the remark this answers is not in the document]
Comment (yinghai) [reply docs/ONNXIFI.md]: Yes. Because we don't spawn a thread just to communicate to backend.

File: docs/ONNXIFI.md [hunk 2]
```diff
@@ -20,13 +20,19 @@ We propose a cross-platform API for loading and executing ONNX graphs on optimiz
...
+8. Call `onnxRunGraph` with the initialized `inputFence` and `outputFence` structures to enable execution of the graph. The call to `onnxRunGraph` will populate `event` member of the `outputFence` with a newly created event object, asynchronously execute the graph once `inputFence`'s `event` is signalled, and then signal the `outputFence`'s `event`.
+9. Call `onnxSignalEvent` with `event` member of `inputFence` to signal to the backend that the inputs are ready to be consumed. 
+10. Call `onnxWaitEvent` (alternatively, repeatedly call `onnxGetEventState` in a loop until the event state is `ONNXIFI_EVENT_STATE_SIGNALLED`) with `event` member of `outputFence` to wait until graph outputs are ready to be consumed. Release events for inputs and outputs using `onnxReleaseEvent`.
+11. If your model works with fixed-size inputs and outputs, or shape and location of inputs and outputs does not change, one call to `onnxSetGraphIO` is sufficient for multiple `onnxRunGraph` calls. For models with variable-size inputs or outputs, you'd need to call `onnxSetGraphIO` before each `onnxRunGraph` call.
```
Comment (rdzhabarov) [inline docs/ONNXIFI.md]: how can you run `onnxRunGraph` multiple times without `onnxSetGraphIO`? I assume you can only do that **if** previous run is completed. Can you mention that there is no expected concurrent execution of `onnxRunGraph` with the same `setGraphIO`?

Comment (zrphercule) [reply docs/ONNXIFI.md]: @rdzhabarov I think onnxSetGraphIO only gives the IO info to backend, therefore it is possible to onnxRunGraph for multiple times but onnxSetGraphIO for only one time? Like
onnxSetGraphIO(...);
//first time running
onnxRunGraph(...);
onnxSignalEvent(...);
onnxWaitEvent(...);
... //get the output, maybe even modified the input descriptor...
//then run a second time without changing the shape and address of input&output
onnxRunGraph(...);
...

File: docs/ONNXIFI.md (hunk 1 above)
[the remark this answers is not in the document]
Comment (Maratyszcza) [reply docs/ONNXIFI.md]: No, the backend should spawn a thread to do the inference, if it does inference on the CPU, and wait until the event is signaled in this thread. If it does inference on something else then the CPU, it instead should to submit a "wait for event" command to the command queue for the async device, e.g. [cudaStreamWaitEvent](https://www.cs.cmu.edu/afs/cs/academic/class/15668-s11/www/cuda-doc/html/group__CUDART__STREAM_gfe68d207dc965685d92d3f03d77b0876.html), [clEnqueueWaitForEvents](https://www.khronos.org/registry/OpenCL/sdk/1.1/docs/man/xhtml/clEnqueueWaitForEvents.html), or [glWaitSync](https://www.khronos.org/registry/OpenGL-Refpages/es3.0/html/glWaitSync.xhtml).

**Pull request merged**